### ISSUE TITLE
Tpetra::Distributor::post() with rank-2 view arguments

### ISSUE THREAD

**Issue by dalg24** [trilinos/Trilinos#2143]
@trilinos/tpetra

## Expectations

I have been using `Tetra::Distributor::doPostsAndWaits()` with rank-1 views and `numPackets = 1` for a while but I am having issues to communicate data from rank-2 views.

I have been able to loop over the 2nd dimension and make temporary copies, something along these lines:
```C++
Kokkos::View<T **> exports;
Kokkos::View<T **> imports;
Tpetra::Distributor distributor( comm );

// allocate memory and fill rank-2 views
// setup distributor

 for ( int d = 0; d < DIM; ++d )
{
  Kokkos::View<T *> tmp_exports( exports.label(), exports.extent( 0 ) );
  Kokkos::deep_copy( tmp_exports, Kokkos::subview( exports, Kokkos::ALL, d ) );
  Kokkos::View<T *> tmp_imports( imports.label(), imports.extent( 0 ) );
  distributor.doPostsAndWaits( tmp_exports, 1, tmp_imports );
  Kokkos::deep_copy( Kokkos::subview( imports, Kokkos::ALL, d ), tmp_imports );
}
```

But I was hoping I would be able to do directly:
```C++
distributor.doPostsAndWaits( exports, DIM, imports );
```

## Current Behavior

`doPostsAndWaits()` does not compile with rank-2 views even though it was not obvious to me it would not from looking at the interface.  The problem is in `Tpetra_Distributor.hpp` in the definition of `doPosts( ExpView, size_t, ImpView )` which has a call to `Kokkos::Compat::subview_offset()` that fails a `static_assert` that the rank of the view is 1.

## Motivation and Context

The workaround with the loop over the 2nd dimension far from ideal.  I would appreciate it if you can point me into the right direction.

## Your Environment

- **Relevant repo SHA1s:**  both the current tip of the develop branch and release 12.12.1 have the issue
- **Relevant configure flags or configure script:**  
- **Operating system and version:**  
- **Compiler and TPL versions:**  

## Related Issues

* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of

**Comment by aprokop** [comment]
@trilinos/tpetra

**Comment by mhoemmen** [comment]
@dalg24 Thanks for your interest!

`Tpetra::Distributor` was meant for communicating packed, one-dimensional arrays.  (I used the passive voice because I didn't come up with Tpetra's original class design.  That came straight from Epetra, long before I started working on Trilinos.  You may replace `Tpetra::Distributor` with `Epetra_*Distributor` and `Tpetra::DistObject` with `Epetra_DistObject` in this comment, and it would remain accurate.)  Users were not really meant to interact with Distributor directly.  This is why it has a lower-level interface that does less error checking.  "Packed" here means "contiguous entries, grouped by MPI process rank."

Users are supposed to use Distributor through DistObject and its subclasses.  Users who want to implement new distributed objects should write a subclass of DistObject.  You've already done most of the work for that, by packing and unpacking data.  DistObject implements two interfaces for backwards compatibility reasons; since you've committed to use Kokkos::View, you should implement the "new" interface.  The header file and its documentation should make this clear. 

Please feel free to post further questions; thanks!

**Comment by dalg24** [comment]
> Tpetra::Distributor was meant for communicating packed, one-dimensional arrays.

OK.  Thanks.

**Comment by mhoemmen** [comment]
@dalg24 Check out `Tpetra::Details::CooMatrix` in `tpetra/core/inout/Tpetra_Details_CooMatrix.hpp` for an example.  That's a sparse matrix in coordinate format, but it illustrates the "new" DistObject interface.  You'll need to implement the following methods:

  - constantNumberOfPackets
  - checkSizes
  - useNewInterface (it should return `true` in your case)
  - copyAndPermuteNew
  - packAndPrepareNew
  - unpackAndCombineNew

Once you do that, you'll be able to use instances of your object with Import and Export operations, just like other Tpetra objects.  You may even decide to accept source objects for Import / Export of different types -- e.g., you might like to Import / Export from a MultiVector into an instance of your type of object.  It's up to you.
